### PR TITLE
Add empty closure lint

### PR DESCRIPTION
**Pull request by ThibsG**
This PR adds a lint that triggers spawned empty closures.

Fixes: #625

changelog: Add empty closure lint

### CODE REVIEW

File: clippy_lints/src/empty_closure.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+use crate::utils::{match_def_path, paths, span_lint};
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for empty closure.
```
Comment (Centril) [inline clippy_lints/src/empty_closure.rs]:
```suggestion
    /// **What it does:** Checks for empty closure when calling `std::thread::spawn`.
```

**Comment by Manishearth** [comment]
I'm not 100% convinced this is a necessary lint, do people ever write this?

File: clippy_lints/src/empty_closure.rs (hunk 1 above)
Comment (Manishearth) [reply clippy_lints/src/empty_closure.rs]: This still needs to be addressed

File: clippy_lints/src/empty_closure.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /// **Why is this bad?** Empty closure does nothing, it can be removed.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// std::thread::spawn(|| {});
+    /// ```
+    pub EMPTY_CLOSURE,
```
Comment (Manishearth) [inline clippy_lints/src/empty_closure.rs]: the name should probably be spawn_empty_closure

File: clippy_lints/src/empty_closure.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    style,
+    "closure with empty body"
```
Comment (Manishearth) [inline clippy_lints/src/empty_closure.rs]: this should probably mention spawning

**Comment by ThibsG** [comment]
> I'm not 100% convinced this is a necessary lint, do people ever write this?

Me neither. I guess people don't (or they don't read their code after refactoring).

Anyway if people are not convinced, we should just close the issue and this PR, it's totally fine.

**Comment by Manishearth** [comment]
Sounds good. Let me know if you want to find something else to work on. We do have a couple issues like this where we filed stuff but weren't yet sure if we wanted them, sorry about that.

**Pull request closed without merging**